include/seqan3/range/view/deep.hpp

```diff
@@ -97,72 +97,135 @@ namespace seqan3::view
 {
 private:
-    //!\brief Hold the inner functor by value or reference depending on deduced type.
-    underlying_adaptor_t inner_functor;
+    //!\brief Type of the CRTP-base.
+    using base_type = detail::adaptor_base<deep<underlying_adaptor_t>, underlying_adaptor_t>;
 
-public:
+    //!\brief Befriend the base class so it can call impl().
+    friend base_type;
 
+public:
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    constexpr deep() = default;
-    constexpr deep(deep const &) = default;
-    constexpr deep(deep &&) = default;
-    constexpr deep & operator=(deep const &) = default;
-    constexpr deep & operator=(deep &&) = default;
-    ~deep() = default;
-
-    //!\brief Constructor that takes the underlying adaptor as parameter.
-    constexpr deep(underlying_adaptor_t && inner)
-        : inner_functor{std::forward<underlying_adaptor_t>(inner)}
-    {}
+    using base_type::base_type;
     //!\}
 
     //!\privatesection
 
-    /*!\brief       The implementation that the base classes operators all resolve to.
+    //!\brief Explicitly inherit operator() because we are also defining our own.
+    using base_type::operator();
+
+    /*!\brief       Unwrap the internal adaptor closure object and pipe the range into it.
+     * \tparam      urng_t Type of the underlying range.
+     * \tparam      underlying_adaptor_t_ Same as underlying_adaptor_t with possibly different cref qualification.
+     * \param[in]   urange The view's underlying range.
+     * \param[in]   adap The stored adaptor unwrapped by the base-classes explode()-member.
+     * \returns     A view with the inner adaptor applied on the innermost ranges.
+     */
+    template <std::ranges::InputRange urng_t, typename underlying_adaptor_t_>
+    static constexpr auto impl(urng_t && urange, underlying_adaptor_t_ && adap)
+    {
+        return std::forward<urng_t>(urange) | std::forward<underlying_adaptor_t_>(adap);
+    }
+
+    /*!\brief       Specialisation of the range-handling `operator()` for range-of-range (this is where deep
+     *              changes the behaviour for nested ranges).
      * \tparam      urng_t Type of the underlying range.
-     * \tparam      arg_types The argument types, not that **temporaries will be copied on recursion!**
      * \param[in]   urange The view's underlying range.
-     * \param[in]   args Further arguments (optional).
      * \returns     A view with the inner adaptor applied on the innermost ranges.
      *
      * \details
      *
      * Recurses and calls std::view::transform if the underlying range is a range-of-ranges.
      */
+    template <std::ranges::InputRange urng_t>
+    //!\cond
+        requires std::ranges::InputRange<reference_t<urng_t>>
+    //!\endcond
+    constexpr auto operator()(urng_t && urange) &
+    {
+        return std::forward<urng_t>(urange) | std::view::transform([me = *this] (auto && e)
+        {
+            return std::forward<decltype(e)>(e) | me;
```

Comment:
ok then I did not understand the exact behaviour.
```
template <typename t>
void foo (t && v)
{
    // within the scope of this function v behaves like an lvalue unless explicitily forwarded
    // so I thought: t != decltype(v)
    // as in t -> int && while decltype(v) -> int &
    // and that's why I need std::move(v) which casts the type back to && ... 
}
```

Reply:
* v is always in lvalue, remember that rvalue references are lvalues, too.
* in your example 
```cpp
int i = 4;
f(i); // t == int & and decltype(v) also == int &

f(3); // t == int and decltype(v)  == int &&

// there are no other combinations not considering const
``` 

but independent of the type, if you call `bar(v)` inside foo, it will always pass `int &` unless you explicitly cast `&&` onto the type again with std::forward. In the case where the type already contains `&` casting `&&` onto it has no effect because of reference collapsing.